include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -61,57 +61,91 @@ template<typename T NBL_STRUCT_CONSTRAINABLE>
-{
-	using return_t = T;
-	static return_t __call(NBL_CONST_REF_ARG(T) val, NBL_CONST_REF_ARG(U) min, NBL_CONST_REF_ARG(U) max)
-	{
-		return spirv::fClamp(val, min, max);
-	}
-};
-template<typename T, typename U> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::uClamp<T>(experimental::declval<T>(), experimental::declval<U>(), experimental::declval<U>()))>)
-struct clamp_helper<T, U NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::uClamp<T>(experimental::declval<T>(), experimental::declval<U>(), experimental::declval<U>()))>) >
-{
-	using return_t = T;
-	static return_t __call(NBL_CONST_REF_ARG(T) val, NBL_CONST_REF_ARG(U) min, NBL_CONST_REF_ARG(U) max)
-	{
-		return spirv::uClamp(val, min, max);
-	}
-};
-template<typename T, typename U> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::sClamp<T>(experimental::declval<T>(), experimental::declval<U>(), experimental::declval<U>()))>)
-struct clamp_helper<T, U NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::sClamp<T>(experimental::declval<T>(), experimental::declval<U>(), experimental::declval<U>()))>) >
-{
-	using return_t = T;
-	static return_t __call(NBL_CONST_REF_ARG(T) val, NBL_CONST_REF_ARG(U) min, NBL_CONST_REF_ARG(U) max)
-	{
-		return spirv::sClamp(val, min, max);
-	}
-};
 
-#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(HELPER_NAME, SPIRV_FUNCTION_NAME, RETURN_TYPE)\
-template<typename T> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>)\
-struct HELPER_NAME<T NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>) >\
+// it is crucial these partial specializations appear first because thats what makes the helpers match SPIR-V intrinsics first
+
+#define DECLVAL(r,data,i,_T) BOOST_PP_COMMA_IF(BOOST_PP_NOT_EQUAL(i,0)) experimental::declval<_T>()
+#define DECL_ARG(r,data,i,_T) BOOST_PP_COMMA_IF(BOOST_PP_NOT_EQUAL(i,0)) const _T arg##i
+#define WRAP(r,data,i,_T) BOOST_PP_COMMA_IF(BOOST_PP_NOT_EQUAL(i,0)) _T
+#define ARG(r,data,i,_T) BOOST_PP_COMMA_IF(BOOST_PP_NOT_EQUAL(i,0)) arg##i
+
+// the template<> needs to be written ourselves
+// return type is __VA_ARGS__ to protect against `,` in templated return types
+#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(HELPER_NAME, SPIRV_FUNCTION_NAME, ARG_TYPE_LIST, ARG_TYPE_SET, ...)\
+NBL_PARTIAL_REQ_TOP(is_same_v<decltype(spirv::SPIRV_FUNCTION_NAME<T>(BOOST_PP_SEQ_FOR_EACH_I(DECLVAL, _, ARG_TYPE_SET))), __VA_ARGS__ >) \
+struct HELPER_NAME<BOOST_PP_SEQ_FOR_EACH_I(WRAP, _, ARG_TYPE_LIST) NBL_PARTIAL_REQ_BOT(is_same_v<decltype(spirv::SPIRV_FUNCTION_NAME<T>(BOOST_PP_SEQ_FOR_EACH_I(DECLVAL, _, ARG_TYPE_SET))), __VA_ARGS__ >) >\
 {\
-	using return_t = RETURN_TYPE;\
-	static inline return_t __call(const T arg)\
+	using return_t = __VA_ARGS__;\
+	static inline return_t __call( BOOST_PP_SEQ_FOR_EACH_I(DECL_ARG, _, ARG_TYPE_SET) )\
 	{\
-		return spirv::SPIRV_FUNCTION_NAME<T>(arg);\
+		return spirv::SPIRV_FUNCTION_NAME<T>( BOOST_PP_SEQ_FOR_EACH_I(ARG, _, ARG_TYPE_SET) );\
```

Comment:
you should fill out the `<>` using `ARG_TYPE_SET` as well not just `<T>`

Reply:
just how your `AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER` in the c++ section is doing
<https://github.com/Devsh-Graphics-Programming/Nabla/blob/644c1dbd4f5e4e12f1ba2497b33d7b8c26355778/include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl#L239>